### PR TITLE
[SPARK-17849] [SQL] Fix NPE problem when using grouping sets

### PR DESCRIPTION
**Pull request by yangw1234**
## What changes were proposed in this pull request?

Prior this pr, the following code would cause an NPE:
`case class point(a:String, b:String, c:String, d: Int)`

`val data = Seq(
point("1","2","3", 1),
point("4","5","6", 1),
point("7","8","9", 1)
)`
`sc.parallelize(data).toDF().registerTempTable("table")`
`spark.sql("select a, b, c, count(d) from table group by a, b, c GROUPING SETS ((a)) ").show()`

The reason is that when the grouping_id() behavior was changed in #10677, some code (which should be changed) was left out. 

Take the above code for example, prior #10677, the bit mask for set "(a)" was `001`, while after #10677 the bit mask was changed to `011`. However, the `nonNullBitmask` was not changed accordingly.

This pr will fix this problem.
## How was this patch tested?

add integration tests

### CODE REVIEW

**Comment by yangw1234** [comment]
cc @davies Would you help reviewing this?

**Comment by yangw1234** [comment]
also cc @hvanhovell

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala [hunk 1]
```diff
@@ -2189,6 +2189,24 @@ class SQLQuerySuite extends QueryTest with SharedSQLContext {
     }
   }
 
+  test("SPARK-17849: grouping set throws NPE") {
```
Comment (rxin) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala]: maybe we can move this into SQLQueryTestSuite, by creating a new grouping_set.q file??

Comment (yangw1234) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala]: @rxin done

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 2]
```diff
@@ -298,10 +298,11 @@ class Analyzer(
           case other => Alias(other, other.toString)()
         }
 
-        val nonNullBitmask = x.bitmasks.reduce(_ & _)
+        val nonNullBitmask = ~ x.bitmasks.reduce(_ | _)
```
Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: +1

Comment (yangw1234) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: Ok, I'll do it.

Comment (yangw1234) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: @hvanhovell @rxin comments are added

**Comment by davies** [comment]
@yangw1234 Thanks for working on this, could you also double check that all the places that use bitmasks are correct?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 3]
```diff
@@ -298,10 +298,14 @@ class Analyzer(
           case other => Alias(other, other.toString)()
         }
 
-        val nonNullBitmask = x.bitmasks.reduce(_ & _)
+        // The left most bit in the bitmasks corresponds to the last expression in groupByAliases
+        // with 0 indicating this expression is in the grouping set. The following line of code
+        // calculates the bit mask representing the expressions that exist in all the grouping sets.
+        val nonNullBitmask = ~ x.bitmasks.reduce(_ | _)
```
Comment (yangw1234) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: Do you mean `((nonNullBitmask >> (attrLength - idx - 1)) & 1) == 1`? We can only test on `0` if we left shift `1`, right? @davies

**Comment by yangw1234** [comment]
@davies Other places all seem to be correct.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 4]
```diff
@@ -298,10 +298,14 @@ class Analyzer(
           case other => Alias(other, other.toString)()
         }
 
-        val nonNullBitmask = x.bitmasks.reduce(_ & _)
+        // The rightmost bit in the bitmasks corresponds to the last expression in groupByAliases with 0
+        // indicating this expression is in the grouping set. The following line of code calculates the
+        // bitmask representing the expressions that exist in all the grouping sets (also indicated by 0).
+        val nonNullBitmask = x.bitmasks.reduce(_ | _)
```
Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: Should we call this `nullBitmask` now? (1 means it's nullable)

**Comment by rxin** [comment]
@yangw1234 can you fix the scala styles?

**Comment by yangw1234** [comment]
scala style fixed.
 I didn't notice. Sorry for the delay. @rxin

**Comment by hvanhovell** [comment]
retest this please

**Comment by davies** [comment]
LGTM, pending tests

**Comment by hvanhovell** [comment]
retest this please

**Comment by hvanhovell** [comment]
I'll merge after a successfull test run

**Comment by hvanhovell** [comment]
LGTM - Merging to master/2.1/2.0. Thanks!

**Pull request closed without merging**